src/optimize/dynamic_dll_plugin/dynamic_dll_plugin.js

```diff
@@ -456,12 +291,43 @@ export class DynamicDllPlugin {
     return stubModule;
   }
 
+  async areMaxCompilationsPerformed() {
+    // Logic to run the max compilation requirements.
+    // Only enable this for CI builds in order to ensure
+    // we have an healthy dll ecosystem.
+    if (this.performedCompilations === this.maxCompilations) {
+      return Promise.reject('All the allowed dll compilations were already performed and one more is needed which is not possible');
+    }
+
+    return Promise.resolve();
+  }
+
   async runDLLCompiler(mainCompiler) {
-    await this.dllCompiler.run(this.entryPaths);
+    const runCompilerErrors = [];
+
+    try {
+      await this.dllCompiler.run(this.entryPaths);
+    } catch (e) {
+      runCompilerErrors.push(e);
+    }
+
+    try {
+      await this.areMaxCompilationsPerformed();
+    } catch (e) {
+      runCompilerErrors.push(e);
+    }
 
     // We need to purge the cache into the inputFileSystem
     // for every single built in previous compilation
     // that we rely in next ones.
     mainCompiler.inputFileSystem.purge(this.dllCompiler.getManifestPath());
+
+    this.performedCompilations++;
+
+    if (!runCompilerErrors.length) {
+      return Promise.resolve();
+    }
+
+    return Promise.reject(runCompilerErrors.join('\n-'));
```

Comment:
This is another case of rejecting/throwing a string, which is problematic because most everybody expects thrown values to be errors. How about `throw new Error(runCompilerErrors.join('\n-'))`. Also, do you know that this formats the different types of errors `dllCompiler.run()` can produce well? What about errors thrown by webpack?

Reply:
👍 throwing right now. Yes this way of doing the things is also compatible with webpack errors 😃